Pack추천알고리즘

hhseo9519 opened the pull request:
### 📌 PR 타입
-[O] 기능 추가
-[ ] 기능 삭제
-[ ] 버그 수정
-[ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트
### ✈️ 반영 브랜치
<!--feat/login -> dev-->

### 📄 관련 이슈
<!-- 이슈번호: #334 -->

### 🧑‍💻 구현한 내용
팩 추천 알고리즘을 통해 3개의 팩을 뽑아내는 기능을 구현했습니다

### 🧪 테스트 결과

### 👿 트러블 슈팅

### 💬 코멘트





<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Endpoint to fetch today’s top 3 recommended starter packs.
  * Packs can include and list individual pack items.

* **Performance**
  * Added database indexing and query optimizations to speed time-based engagement lookups and recommendation calculations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

hhseo9519 committed:
feat : Pack 추천 dto

hhseo9519 committed:
feat : Pack 추천 service

hhseo9519 committed:
feat : index 추가

hhseo9519 committed:
feat : pack 추천 repository

hhseo9519 committed:
feat : table에 index 추가

hhseo9519 committed:
feat : pack 추천 api 작성

hhseo9519 committed:
feat : sql 변경

hhseo9519 committed:
Merge branch 'develop' into pack추천알고리즘

Lee-seung-won commented on src/main/java/com/starterpack/pack/service/PackRecommendService.java:
```diff
@@ -0,0 +1,194 @@
+            );
+
+            feats.add(new F(p, Ltotal, L24h, L7d, Lprev24h, streakDays, hoursSinceLastLike, likeTimes));
+        }
```
n+1문제가 발생할 것 같은데 한번 확인해주세요!

hhseo9519 committed:
feat : N+1 문제 해결

# Conflicts:
#	src/main/java/com/starterpack/pack/repository/PackRepository.java

hhseo9519 committed:
feat : LocalDateTime으로 바로 캐스팅하면 생기는 ClassCastException 해결

Lee-seung-won commented on src/main/java/com/starterpack/pack/service/PackRecommendService.java:
```diff
@@ -0,0 +1,246 @@
+    private static final double HALF_LIFE_HOURS = 24.0;  // 감쇠 반감기(시간)
+
+    public List<PackRecommendDto> getTodayTop3() {
+        List<Pack> packs = packRepository.findAll();
```
여기서 findAll메서드는 items,category, member 모두 fetch join하는데 추천 메서드는 Pack의 id와 packLikeCount만 사용하는 것으로 보여요 ! 그래서 불필요하게 fetch join을 수행해서 팩의 개수가 500개가 넘어가면 서버가 다운되는 등의 문제가 발생할 수 있다고 합니다 ! 아래와 같이 추천 메서드 전용 find메서드를 만들어서 사용하는게 어떨까요?

```java
// PackRepository에 추가
@Query("select p from Pack p")
List<Pack> findAllForRecommendation();
```

Lee-seung-won reviewed (approved):
고생하셨습니다! 코멘트 하나 달았는데 그것만 확인하고 머지 하셔도 될거 같아요 !
